drivers/epd_bw_spi/epd_bw_spi.c

```diff
@@ -0,0 +1,329 @@
+/*
+ * Copyright (C) 2019 Silke Hofstra
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_epd_bw_spi
+ *
+ * @{
+ * @file
+ * @brief       Device driver implementation for the epd_bw_spi display controller
+ *
+ * @author      Silke Hofstra <silke@slxh.eu>
+ */
+#include <string.h>
+#include "byteorder.h"
+#include "epd_bw_spi.h"
+
+#include "epd_bw_spi.h"
+#include "epd_bw_spi_params.h"
+#include "epd_bw_spi_internal.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+void epd_bw_spi_cmd_start(epd_bw_spi_params_t *p, uint8_t cmd, bool cont)
+{
+    DEBUG("[epd_bw_spi] cmd_start: command 0x%02x\n", cmd);
+    if (p->busy_pin != GPIO_UNDEF) {
+        while (gpio_read(p->busy_pin) == p->busy_value) {}
+    }
+    gpio_clear(p->dc_pin);
+    spi_transfer_byte(p->spi, p->cs_pin, cont, (uint8_t)cmd);
+    gpio_set(p->dc_pin);
+}
+
+void epd_bw_spi_write_cmd(epd_bw_spi_params_t *p, uint8_t cmd,
+                          const uint8_t *params, size_t plen)
+{
+    spi_acquire(p->spi, p->cs_pin, SPI_MODE_0, p->spi_clk);
+    epd_bw_spi_cmd_start(p, cmd, plen > 0);
+    if (plen) {
+        spi_transfer_bytes(p->spi, p->cs_pin, false, params, NULL, plen);
+    }
+    spi_release(p->spi);
+}
+
+void epd_bw_spi_read_cmd(epd_bw_spi_params_t *p, uint8_t cmd, uint8_t *params,
+                         size_t plen)
+{
+    spi_acquire(p->spi, p->cs_pin, SPI_MODE_0, p->spi_clk);
+    epd_bw_spi_cmd_start(p, cmd, true);
+    spi_transfer_bytes(p->spi, p->cs_pin, false, NULL, params, plen);
+    spi_release(p->spi);
+}
+
+void epd_bw_spi_wait(epd_bw_spi_params_t *p, uint32_t usec)
+{
+    if (p->busy_pin != GPIO_UNDEF) {
+        DEBUG("[epd_bw_spi] wait: for busy bin\n");
+        while (gpio_read(p->busy_pin)) {}
+    }
+    else {
+        DEBUG("[epd_bw_spi] wait: for %" PRIu32 " microseconds\n", usec);
+        xtimer_usleep(usec);
+    }
+}
+
+int epd_bw_spi_init(epd_bw_spi_t *dev,
+                    const epd_bw_spi_params_t *params,
+                    const epd_bw_spi_controller_t *controller,
+                    uint8_t size_x, uint16_t size_y,
```

Comment:
size_x and size_y are set via the params (and there is a mismatch between values set in the different controller structs and these values). Same for the controller, there could be one `epd_bw_spi_controller_t` attribute in the params.
Then the signature of the `epd_bw_spi_init` function would be consistent with most of other drivers: `
int epd_bw_spi_init(epd_bw_spi_t *dev, const epd_bw_spi_params_t *params);`.

Reply:
I described my idea for this in [this comment](https://github.com/RIOT-OS/RIOT/pull/12509#issuecomment-826296194).

Reply:
I see, this looks nicer but not from the params struct static initializer point of view. I think a flat struct is ok here:

```c
/**
 * @brief   Driver parameters.
 */
typedef struct {
    spi_t spi;          /**< SPI device that the display is connected to */
    spi_clk_t spi_clk;  /**<; SPI clock speed to use */
    gpio_t cs_pin;      /**<; pin connected to the CHIP SELECT line */
    gpio_t dc_pin;      /**< pin connected to the DC line */
    gpio_t rst_pin;     /**< pin connected to the reset line */
    gpio_t busy_pin;    /**< pin connected to the busy line */
    bool busy_value;    /**< expected value for the busy pin */
    bool dummy;         /**< if device requires a dummy cycle before read */
    epd_bw_spi_entry_mode_t entry_mode;
    uint16_t size_x;    /**< supported number of horizontal pixels */
    uint16_t size_y;    /**< supported number of vertical pixels */
    epd_bw_spi_controller_t controller; /**< E-paper controller configuration */
} epd_bw_spi_params_t;
```

And then in the params:

```c
/**
 * @brief    Default controller definition
 */      
#ifndef EPD_BW_SPI_CONTROLLER
#define EPD_BW_SPI_CONTROLLER EPD_BW_SPI_CONTROLLER_IL3829
#endif

#define EPD_BW_SPI_PARAMS           { .spi = EPD_BW_SPI_PARAM_SPI, \
                                          .spi_clk = EPD_BW_SPI_PARAM_SPI_CLK, \
                                          .cs_pin = EPD_BW_SPI_PARAM_CS, \
                                          .dc_pin = EPD_BW_SPI_PARAM_DC, \
                                          .rst_pin = EPD_BW_SPI_PARAM_RST, \
                                          .busy_pin = EPD_BW_SPI_PARAM_BUSY, \      
                                          .busy_value = EPD_BW_SPI_PARAM_BUSY_VAL, \
                                          .dummy = EPD_BW_SPI_PARAM_DUMMY, \
                                          .entry_mode = EPD_BW_SPI_PARAM_ENTRY_MODE, \
                                          .size_x = EPD_BW_SPI_PARAM_DISPLAY_X, \
                                          .size_y = EPD_BW_SPI_PARAM_DISPLAY_Y, \
                                          .controller = EPD_BW_SPI_PARAM_CONTROLLER, \
}
```

Reply:
Ok, you can keep the X/Y size hard coded for each controller, but X/Y attributes are still needed in the params struct, as showed above. In the code, a sanity check/assert could be added (maybe there's one already, I haven't checked!) to verify that the size_x/size_y configured in the params are within the bounds supported by the controller.